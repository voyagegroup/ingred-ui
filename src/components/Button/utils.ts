import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
