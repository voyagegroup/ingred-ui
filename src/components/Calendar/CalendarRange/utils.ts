import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
