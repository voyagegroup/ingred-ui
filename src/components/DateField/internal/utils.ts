export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
