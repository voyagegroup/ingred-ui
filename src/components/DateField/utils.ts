import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
