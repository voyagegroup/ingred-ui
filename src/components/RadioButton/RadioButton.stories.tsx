import React from "react";
import { Story } from "@storybook/react/types-6-0";
import RadioButton, { RadioButtonProps, RadioButtonSize } from "./";

export default {
  title: "Components/Inputs/RadioButton",
  component: RadioButton,
  args: {
    size: RadioButtonSize.MEDIUM,
  },
};

export const Example: Story<RadioButtonProps> = (args) => (
  <>
    <RadioButton {...args}>One</RadioButton>
    <br />
    <RadioButton {...args}>Two</RadioButton>
  </>
);

Example.args = {
  name: "example",
};

export const DesignSamples = () => (
  <>
    <RadioButton>Basic</RadioButton>
    <br />
    <RadioButton checked={true}>Selected</RadioButton>
    <br />
    <RadioButton disabled={true}>Disabled</RadioButton>
    <br />
    <RadioButton disabled={true} checked={true}>
      Disabled(checked)
    </RadioButton>
  </>
);
