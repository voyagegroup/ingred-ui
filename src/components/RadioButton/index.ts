export { default, RadioButtonSize } from "./RadioButton";
export type { RadioButtonProps } from "./RadioButton";
