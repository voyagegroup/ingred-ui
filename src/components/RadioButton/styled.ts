import styled, { css } from "styled-components";
import { fontSize } from "../Typography/Typography";
import { RadioButtonSize } from "./RadioButton";

export const Wrapper = styled.label<{
  size: RadioButtonSize;
  disabled?: boolean;
}>`
  position: relative;
  display: inline-flex;
  align-items: center;
  gap: ${({ theme }) => theme.spacing / 2}px;
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
  font-size: ${({ size }) =>
    size === RadioButtonSize.SMALL ? `${fontSize.sm}px` : `${fontSize.md}px`};

  ${({ disabled }) =>
    disabled &&
    css`
      color: ${({ theme }) => theme.palette.text.disabled};
    `}

  & > input {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
    aspect-ratio: 1;
    width: ${({ size }) => size};
    height: auto;
    opacity: 0;
    cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
  }
`;

type IndicatorProps = {
  size: RadioButtonSize;
  inside: string;
  border: string;
};

export const Indicator = styled.span<IndicatorProps>`
  position: relative;
  display: block;
  flex: 0 0 auto;
  aspect-ratio: 1;
  width: ${({ size }) => size};
  height: auto;
  border-radius: 50%;
  border: ${({ border }) => border} solid
    ${({ theme }) => theme.palette.divider};
  box-shadow: ${({ theme }) => theme.shadow["3dShadowBasic"]};
  transition: all 0.3s ease;
  background: ${({ theme }) => theme.palette.background.default};

  input:checked + & {
    background: ${({ theme }) => theme.palette.primary.main};
    border-color: ${({ theme }) => theme.palette.primary.dark};
    box-shadow: ${({ theme }) => theme.shadow["3dShadowPrimary"]};
  }

  input:disabled + & {
    background: ${({ theme }) => theme.palette.gray.light};
    border-color: ${({ theme }) => theme.palette.text.disabled};
  }

  // 選択時の中心の○部分
  &::after {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: block;
    margin: auto;
    border-radius: 50%;
    content: "";
    aspect-ratio: 1;
    width: ${({ inside }) => inside};
    height: auto;
    background: ${({ theme }) => theme.palette.background.default};
    transition-property: background;
  }

  input:disabled + &::after {
    background: ${({ theme }) => theme.palette.gray.light};
  }

  input:disabled:checked + &:after {
    background: ${({ theme }) => theme.palette.text.disabled};
  }
`;

export const Label = styled.span`
  flex: 0 1 auto;
  font-size: inherit;
  color: inherit;
`;
