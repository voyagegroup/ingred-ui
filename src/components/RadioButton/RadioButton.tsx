import * as React from "react";
import styled from "styled-components";
import { hexToRgba } from "../../utils/hexToRgba";

export enum RadioButtonSize {
  // MEDIUM = "24px",
  MEDIUM = "18px",
}

const Wrapper = styled.label<{ size: RadioButtonSize; disabled?: boolean }>`
  position: relative;
  display: inline-flex;
  align-items: center;
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};

  & > input {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;
    width: ${({ size }) => size};
    height: ${({ size }) => size};
    opacity: 0;
    cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
  }
`;

const indicatorSizes = {
  [RadioButtonSize.MEDIUM]: {
    inside: "8px",
    border: "1px",
  },
};

type IndicatorProps = {
  size: RadioButtonSize;
  inside: string;
  border: string;
};
const Indicator = styled.div<IndicatorProps>`
  position: relative;
  display: block;
  flex: 1 0 auto;
  width: ${({ size }) => size};
  height: ${({ size }) => size};
  border-radius: 50%;
  border: ${({ border }) => border} solid
    ${({ theme }) => theme.palette.divider};
  box-shadow: ${({ theme }) =>
    `0px -2px ${hexToRgba(
      theme.palette.black,
      0.16,
    )} inset, 0px 2px ${hexToRgba(theme.palette.black, 0.08)}`};
  transition: all 0.3s ease;
  background: ${({ theme }) => theme.palette.background.default};

  &::after {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: block;
    margin: auto;
    border-radius: 50%;
    content: "";
  }

  &::after {
    width: ${({ inside }) => inside};
    height: ${({ inside }) => inside};
    background: ${({ theme }) => theme.palette.background.default};
    transition-property: background;
  }

  input:checked + & {
    background: ${({ theme }) => theme.palette.primary.main};
    border-color: ${({ theme }) => theme.palette.primary.dark};
  }

  input:disabled + & {
    background: ${({ theme }) => theme.palette.gray.light};
    border-color: ${({ theme }) => theme.palette.text.disabled};
  }
  input:disabled + &:after {
    background: ${({ theme }) => theme.palette.gray.light};
  }
  input:disabled:checked + &:after {
    background: ${({ theme }) => theme.palette.text.disabled};
  }
`;

export const Label = styled.span`
  flex: 0 1 auto;
  margin-left: ${({ theme }) => theme.spacing / 2}px;
  font-size: 14px;

  input:disabled + div + & {
    color: ${({ theme }) => theme.palette.text.disabled};
  }
`;

export type RadioButtonChangeHandler = (
  e: React.ChangeEvent<HTMLInputElement>,
  checked: boolean,
) => void;

export type RadioButtonProps = Omit<
  React.InputHTMLAttributes<HTMLInputElement>,
  "size" | "onChange"
> & {
  size?: RadioButtonSize;
  children?: React.ReactNode;
  onChange?: RadioButtonChangeHandler;
  inputRef?: React.Ref<HTMLInputElement>;
};

const RadioButton = React.forwardRef<
  HTMLLabelElement | HTMLSpanElement,
  RadioButtonProps
>(
  (
    {
      size = RadioButtonSize.MEDIUM,
      onChange,
      inputRef,
      children,
      disabled,
      ...rest
    },
    ref,
  ) => {
    const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      if (onChange) {
        onChange(e, e.target.checked);
      }
    };
    return (
      <Wrapper
        ref={ref}
        as={children == null ? "span" : "label"}
        disabled={disabled}
        size={size}
      >
        <input
          {...rest}
          ref={inputRef}
          disabled={disabled}
          type="radio"
          onChange={handleChange}
        />
        <Indicator size={size} {...indicatorSizes[size]} />
        {children != null ? <Label>{children}</Label> : null}
      </Wrapper>
    );
  },
);

export default RadioButton;
