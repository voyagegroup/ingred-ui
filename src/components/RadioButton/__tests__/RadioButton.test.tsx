import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import RadioButton from "../RadioButton";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("RadioButton component testing", () => {
  afterEach(cleanup);

  test("RadioButton", () => {
    const { asFragment } = renderWithThemeProvider(
      <RadioButton>RadioButton</RadioButton>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
