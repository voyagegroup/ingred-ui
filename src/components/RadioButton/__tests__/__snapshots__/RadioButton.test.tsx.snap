// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton component testing RadioButton 1`] = `
<DocumentFragment>
  <label
    class="sc-bdvvtL kSfpce"
  >
    <input
      type="radio"
    />
    <div
      class="sc-gsDKAQ fyfwNU"
    />
    <span
      class="sc-dkPtRN bgcwTO"
    >
      RadioButton
    </span>
  </label>
</DocumentFragment>
`;
